vscode-dotnet-runtime-library/src/Utils/NodeIPCMutex.ts

```diff
@@ -0,0 +1,286 @@
+        if (process.platform === 'win32')
+        {
+            return `${ipcPathDir}vscd-${id}-sock`;
+        }
+        else
+        {
+            return path.join(ipcPathDir, `vscd-${id}.sock`);
+        }
+    }
+
+    /**
+     * @remarks This function will try to hold a lock to prevent both other processes and other async code in this processes from running simultaneously..
+     * It will retry acquiring the lock if it is already held by another process or async code in this process.
+     * It will also check if the lock is stale (i.e., if the process holding the lock has died) and clean it up if necessary.
+     * It will fail and throw if it cannot acquire the lock after the specified number of retries.
+     *
+     * @param fn - The function to run while we have the lock. This should be a promise, and will be awaited with its value returned.
+     * Use the () => {} syntax to ensure your scope is correctly bound to the function so it can access the variables you declare.
+     * @param retryDelayMs The number of milliseconds to wait before retrying to acquire the lock if it is already held by another process.
+     * @param timeoutTimeMs The total amount of time to try to acquire the lock before giving up. This is the maximum time to wait for the lock to be released.
+     * @param actionId The action ID to use for logging and debugging purposes. This should be a unique identifier for the action being performed.
+     * @returns The awaited value returned by the function passed in as fn.
+     */
+    public async acquire<T>(fn: () => Promise<T>, retryDelayMs = 100, timeoutTimeMs = 1000, actionId: string): Promise<T>
+    {
+        const maxRetryCountToEndAtRoughlyTimeoutTime = Math.ceil(timeoutTimeMs / retryDelayMs) + 1;
+        let retries = 0;
+
+        while (true)
+        {
+            try
+            {
+                return await this.tryAcquire(actionId, fn);
+            }
+            catch (error: any)
+            {
+                // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+                if (error?.code === 'EADDRINUSE') //  We couldn't acquire the lock, even though nobody else is using it.
+                {
+                    if (retries >= maxRetryCountToEndAtRoughlyTimeoutTime)
+                    {
+                        throw new Error(`Action: ${actionId} Failed to acquire lock after ${retries} retries out of ${maxRetryCountToEndAtRoughlyTimeoutTime} total available attempts.
+The lock: ${this.lockPath} may be held by another process or instance of vscode. Try restarting your machine, deleting the lock, and or increasing the timeout time in the extension settings.
+
+Increase your OS path length limit to at least 256 characters.
+On Linux, you can set XDG_RUNTIME_DIR to be a writeable directory by your user.
+
+If you still face issues, set VSCODE_DOTNET_RUNTIME_DISABLE_MUTEX=true in the environment.
+Report this issue to our vscode-dotnet-runtime GitHub for help.`);
```

Comment:
I think I've asked this before, but we don't do localization for this extension, right?

Reply:
That's correct, I need to add that eventually, though not in this PR. This error message should probably be customizable if it were to be an external library I ported out as well for other tools to have their own error.

This error message is quite long, but the problem is complex and there are many possible solutions/issues. I would be happy to have any suggestions to make it shorter/better.